cleo_sdk/CLEO.h

```diff
@@ -97,42 +97,84 @@ static const char* ToStr(eDataType type)
 	default: return "corrupted";
 	}
 }
+static bool IsInteger(eDataType type)
+{
+	switch (type)
+	{
+		case DT_BYTE:
+		case DT_WORD:
+		case DT_DWORD:
+			return true;
+	}
+	return false;
+}
+static bool IsString(eDataType type)
+{
```

Comment:
Why IsString includes variables and IsInteger does not? IsVariable should be part of IsInteger

Reply:
Because there is no DT_VAR_INT.  
DT_VAR actually can contain multiple types.

Reply:
Maybe if you call it IsNumber then it would better reflect what it does

Reply:
How should it be called then?

Reply:
Make 2 functions: IsNumber and IsString.

If IsNumber returns true, it's a label, one path. If IsString, it's a module, another path. Otherwise, error

Reply:
For CollectNumberParams there is no difference between literal numbers and variables with numbers, sane fir string arguments

Reply:
So, variables should be counted in in both IsNumber and IsString functions? Does it make sense to IsString behave same as (IstString || IsVariable) ?

Reply:
> Make 2 functions: IsNumber and IsString.
> 
> If IsNumber returns true, it's a label, one path. If IsString, it's a module, another path. Otherwise, error

That's how it is done now
https://github.com/cleolibrary/CLEO5/pull/50/files#diff-406042e278a670ad6fc9d7d92af65c44f6db4726c1b8157bc2cb60db595c64edR2149

Reply:
```c++
static bool IsNumber(eDataType type)
{
	switch (type)
	{
		case DT_BYTE:
		case DT_WORD:
		case DT_DWORD:
		case DT_VAR:
		case DT_VAR_ARRAY:
		case DT_LVAR:
		case DT_LVAR_ARRAY:
			return true;
	}
	return false;
}

static bool IsString(eDataType type)
{
	switch (type)
	{
		case DT_STRING:
		case DT_TEXTLABEL:
		case DT_LVAR_TEXTLABEL:
		case DT_LVAR_TEXTLABEL_ARRAY:
		case DT_LVAR_STRING:
		case DT_LVAR_STRING_ARRAY:
		case DT_VAR_TEXTLABEL:
		case DT_VAR_TEXTLABEL_ARRAY:
		case DT_VAR_STRING:
		case DT_VAR_STRING_ARRAY:
		case DT_VARLEN_STRING:
			return true;
	}
	return false;
}



...


if (IsNumber(paramType))
{
	*thread >> label; // label offset
}
else if (IsString(paramType))
{
	moduleTxt = ReadStringParam(thread); // string with module and export name
}
```

Reply:
Um what about floats? Float is number too, but when function is named IsInteger it will lie about variable types caring float.

Reply:
There is no IsInteger in my example

Reply:
Yes, but calling 0AB1 with float as label makes no sense.

Reply:
in the example I gave, both IsNumber and  IsString return false for DT_FLOAT.

Reply:
I would probably add DT_FLOAT to IsNumber btw

```
static bool IsNumber(eDataType type)
{
	switch (type)
	{
		case DT_BYTE:
		case DT_WORD:
		case DT_DWORD:
                case DT_FLOAT:
		case DT_VAR:
		case DT_VAR_ARRAY:
		case DT_LVAR:
		case DT_LVAR_ARRAY:
			return true;
	}
	return false;
}
```

Reply:
you can not guarantee correctness of the given argument based on the data type. DT_VAR can contain float value, DT_DWORD can be way to off to be a valid label, etc. This is just approximation

Reply:
Hm I added float to isNumber and now every time I use this function I do `(isNumber && x != DT_FLOAT)`
Also I would use these functions for example for checking type of 'arguments count' param of 0AB1 (only literal intefers) and it gets complicated too, as I need do another check excluding some types again.

Reply:
Why do we even need those checks? Original opcodes never validate input arguments, they just expect one type or another.

We need number vs string here because we add new functionality to the existing opcode, but that's just one opcode. Where else we need this check?

Reply:
I'm adding checks to most common mistakes, like forgetting to include 'argument count' at all. It cuts all trouble shooting and gives instant answer.

Reply:
ok, how about making this:

```
static bool IsInteger(eDataType type)
{
	switch (type)
	{
		case DT_BYTE:
		case DT_WORD:
		case DT_DWORD:
		case DT_VAR:
		case DT_VAR_ARRAY:
		case DT_LVAR:
		case DT_LVAR_ARRAY:
			return true;
	}
	return false;
}

static bool IsFloat(eDataType type)
{
	switch (type)
	{
                case DT_FLOAT:
		case DT_VAR:
		case DT_VAR_ARRAY:
		case DT_LVAR:
		case DT_LVAR_ARRAY:
			return true;
	}
	return false;
}
```

My point here is that you should not make separate checks for literal values (1, 5, 0, -100.0) and variables (0@, $var). in general case they are interchangeable as opcode argument

Reply:
DT_VAR_ARRAY and DT_LVAR_ARRAY can be further inspected for array type (i or f) if you want to go deeper.

Reply:
Makes sense, but can be more misleading. IsFloat will be true even if variable contains pointer. That's why having `IsInteger || IsVariable` better as it allows to craft any needed condition.
Now each type can be checked by single statement:
* type == DT_FLOAT
* IsInteger
* IsString
* IsVariable
It also forces user to acknowledge the fact that variable can hold whatever.

Reply:
> DT_VAR_ARRAY and DT_LVAR_ARRAY can be further inspected for array type (i or f) if you want to go deeper.

How is it possible? All I know array param is 6 bytes long. This might be useful for current parameter type validation.

Reply:
`IsVariable` tells you nothing about the value.

0@ = 100
0@ = 100.0

are two different values in DT_VAR

Reply:
> IsVariable tells you nothing about the value.

`bool IsVariable(eDataType type)`
it tells me that data type is `variable`

Reply:
> IsInteger || IsVariable in my opinion is better as it allows to craft any needed condition.

```
static bool IsInteger(eDataType type)
{
	switch (type)
	{
		case DT_BYTE:
		case DT_WORD:
		case DT_DWORD:
		case DT_VAR:
		case DT_VAR_ARRAY:
		case DT_LVAR:
		case DT_LVAR_ARRAY:
			return true;
	}
	return false;
}
```

gives you the same result

Reply:
yes, but we are not crafting SDK method to suit one single use case

Reply:
it's not a single use case. 

> My point here is that you should not make separate checks for literal values (1, 5, 0, -100.0) and variables (0@, $var). in general case they are interchangeable as opcode argument

Any time you expect an integer value in opcode you should check for both literal (DT_BYTE, DT_WORD, DT_DWORD) and variable (DT_VAR, DT_VAR_ARRAY, DT_LVAR, DT_LVAR_ARRAY) types

Reply:
But some times I expect literal integer. Then I have to undo what IsInteger did inside. IsVariable tells it can be integer inside, while unified function will tell lie in some cases.

Maybe it should be called `IsConstInteger` `IsLiteralInt` or something like that then?

Reply:
> Any time you expect an integer value in opcode you should check for both literal (DT_BYTE, DT_WORD, DT_DWORD) and variable (DT_VAR, DT_VAR_ARRAY, DT_LVAR, DT_LVAR_ARRAY) types
Yes you are validating parameter TYPE before reading it's VALUE.

Reply:
> But some times I expect literal integer. 

When?

Reply:
> Maybe it should be called `IsConstInteger` `IsLiteralInt` or something like that then?
But then it will be clashing with IsString which returns true for both literals and for-sure string variables.

Rename `IsVariable` to `IsDynamic`? Hm.. also seems not helpful and do not reflects `DT_` typenames.

Reply:
> > But some times I expect literal integer.
> 
> When?

In some cases, like mentioned function `arguments count` which is compile constant too, so makes no sense to be variable.

Reply:
These functions reflect type to text function
```
static const char* ToKindStr(eDataType type)
{
	switch (type)
	{
		case DT_BYTE:
		case DT_WORD:
		case DT_DWORD:
			return "int"; break;

		case DT_FLOAT:
			return "float"; break;

		case DT_STRING:
		case DT_TEXTLABEL:
		case DT_LVAR_TEXTLABEL:
		case DT_LVAR_TEXTLABEL_ARRAY:
		case DT_LVAR_STRING:
		case DT_LVAR_STRING_ARRAY:
		case DT_VAR_TEXTLABEL:
		case DT_VAR_TEXTLABEL_ARRAY:
		case DT_VAR_STRING:
		case DT_VAR_STRING_ARRAY:
		case DT_VARLEN_STRING:
			return "string"; break;

		case DT_VAR:
		case DT_VAR_ARRAY:
		case DT_LVAR:
		case DT_LVAR_ARRAY:
			return "variable"; break;

		case DT_END:
			return "varArgEnd"; break;

		default:
			return "corrupted"; break;
	}
}
```
Motivation was to just make simpler verification of DATA TYPE byte without need for switches, where as already proofed few times, it is to easy to include some cases.

These utilities are meant to verify eDataType sub-category only, in abstraction of actual in code use (as mentioned it would be possible to deduce int-float value type within arrays).

Reply:
I still don't get why we separate `int` and `var` while combine `string` and `stringvar`. should be some consistency at least. we either combine literals and variables (for int, float, and string types) under the same check, or separate them out.

Reply:
As mentioned because there is `DT_LVAR_STRING` but no `DT_LVAR_INT`. I just organized DT_x types together in groups of same flavour.

Reply:
I'm splitting then the functions to separate immediate and variable values.

Reply:
Agreed!